r/R/dplyr.R

```diff
@@ -219,6 +219,31 @@ tail.arrow_dplyr_query <- function(x, n = 6L, ...) {
   x
 }
 
+#' Show the details of an Arrow Execution Plan
+#'
+#' This is a function which gives more details about the Execution Plan (`ExecPlan`)
+#' of an `arrow_dplyr_query` object. It is similar to `dplyr::explain()`.
+#'
+#' @param x an `arrow_dplyr_query` to print the `ExecPlan` for.
+#'
+#' @return The argument, invisibly.
+#' @export
+#'
+#' @examplesIf arrow_with_dataset() & requireNamespace("dplyr", quietly = TRUE)
+#' library(dplyr)
+#' mtcars %>%
+#'   arrow_table() %>%
+#'   filter(mpg > 20) %>%
+#'   mutate(x = gear/carb) %>%
+#'   show_exec_plan()
+show_exec_plan <- function(x) {
+  adq <- as_adq(x)
+  plan <- ExecPlan$create()
+  final_node <- plan$Build(x)
+  cat(plan$ToString())
```

Comment:
You're missing a few things here that get handled in ExecPlan_run: https://github.com/apache/arrow/blob/master/r/src/compute-exec.cpp#L67-L89

So plan$ToString() will miss sorting and sorting + head/tail (topK). If you add some tests that include `arrange()`, you'll see they're not showing up. 

You could add a C++ function that does that first part of ExecPlan_run and creates the sink nodes, then calls ToString(). That would print everything.

Reply:
You are right, if the pipeline includes calls to `arrange()` and / or `head()` they are not captured by `show_exec_plan()`.  Moreover, the filter node gets lost when we use `head()`.

``` r
  mtcars %>% 
    arrow_table() %>% 
    filter(mpg > 20) %>% 
    arrange(desc(wt)) %>% 
    head(3) %>% 
    show_exec_plan()
#> ExecPlan with 2 nodes:
#> 1:ProjectNode{projection=[mpg, cyl, disp, hp, drat, wt, qsec, vs, am, gear, carb]}
#>  0:SourceNode{}

mtcars %>% 
    arrow_table() %>% 
    filter(mpg > 20) %>% 
    arrange(desc(wt)) %>%
    show_exec_plan()
#> ExecPlan with 3 nodes:
#> 2:ProjectNode{projection=[mpg, cyl, disp, hp, drat, wt, qsec, vs, am, gear, carb]}
#>   1:FilterNode{filter=(mpg > 20)}
#>     0:TableSourceNode{}
```

<sup>Created on 2022-07-13 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Let me do some research / reading and will get back on this. It might be beyond my C++ skillset.

Reply:
Interesting about the filter node, I didn't expect that. I wonder what's going on there. (That's one of the great things about adding this feature: who knows what else we'll see in the exec plans that doesn't match what we think it should be.)

Reply:
I think I understand what is going on here. We effectively have 2 ExecPlans because the `order_by_sink` option (involved in `arrange()`) is a _pipeline breaker_ and fully materialises the input into memory (thus splitting the ExecPlan into 2 pars). The first ExecPlan starts with a `TableSourceNode`, includes a `FilterNode` and most likely ends with an `OrderBySinkNode`. The second ExecPlan starts with a `SourceNode` and ends with the `SinkNode` corresponding to `head()`. The print method only captures the second (final) ExecPlan. Maybe we could think how we want to capture situations in which "multiple" ExecPlans are involved (i.e. we have an operation that is a pipeline breaker).

Reply:
On a second pass, it chains the exec plans correctly when there are at least 2 sink nodes (but still, the sink nodes themselves are missing):
``` r
ds %>%
  filter(lgl) %>%
  arrange(chr) %>%
  head() %>%
  show_exec_plan()
#> ExecPlan with 3 nodes:
#> 2:ProjectNode{projection=[int, dbl, lgl, chr, fct, ts, part]}
#>   1:FilterNode{filter=lgl}
#>     0:SourceNode{}
#> ExecPlan with 2 nodes:
#> 1:ProjectNode{projection=[int, dbl, lgl, chr, fct, ts, part]}
#>   0:SourceNode{}
```

<sup>Created on 2022-07-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>